Added Close() to the various go-tspi structs.

DenisKarch opened the pull request:
We typically do not expect to use Close() since our sessions are
relatively short--upon Closing the Context all associated objects
are closed anyway.
Also added GetCapability() to context.

DenisKarch committed:
Added Close() to the various go-tspi structs.
We typically do not expect to use Close() since our sessions are
relatively short--upon Closing the Context all associated objects
are closed anyway.

ericchiang commented on tspi/nv.go:
```diff
@@ -26,8 +26,13 @@ type NV struct {
+	if err != nil {
+		return nil, err
+	}
+	data := C.GoBytes(unsafe.Pointer(cdata), C.int(length))
```
Does this actually create a copy of the cdata?

DenisKarch replied:
Yep, according to [golang.org](https://golang.org/cmd/cgo/) it should

> A few special functions convert between Go and C types by making copies of the data. In pseudo-Go definitions:
```
// C data with explicit length to Go []byte
func C.GoBytes(unsafe.Pointer, C.int) []byte
```

ericchiang commented on tspi/context.go:
```diff
@@ -132,3 +132,11 @@ func (context *Context) CreateHash(hash crypto.Hash) (*Hash, error) {
 }
+
+// GetCapability reads the requested capability and subcapability from the TPM.
+func (context *Context) GetCapability(capa int, subcaplen uint, subcap uint8) ([]byte, error) {
```
Can you open this as a different PR?

Also, I know with TPM 2.0 the TPM isn't actually guaranteed to return all the capabilities on the first call. Do you know what the behavior here is? Can we document that?

Can we add constants for the caps and subcaps?

DenisKarch replied:
The constants should already be in [tspiconst.go](https://github.com/google/go-tspi/blob/master/tspiconst/tspiconst.go).

It looks like all of the TPM 1.2 capabilities are returned as two UINT32 bit maps which are written into a BYTE**. (I can't find any mention as to whether any of the capabilities exceed this size so all the data should be returned per call?)
As far as I can tell the intended use and behaviour is that you call this method for each capability you want--it cannot return multiple capabilities--(if that is what you are asking).

I will move this to a different PR.

DenisKarch committed:
Modified the call to NV_ReadValue.
golang has begun to strongly enforce the
no go-pointer to go-pointer rule thus preventing
the previous version from compiling.
